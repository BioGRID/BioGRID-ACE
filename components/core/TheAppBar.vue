<template>
    <header class="the-app-bar">
        <v-app-bar
            :clipped-left="clipped"
            color="primary"
            app
            elevation="0"
            class="black--text"
            :elevate-on-scroll="false"
            :hide-on-scroll="true"
            :fade-img-on-scroll="true"
        >
            <v-app-bar-nav-icon
                color="tertiary"
                class="mr-3 navDrawerIcon"
                @click.stop="toggleNavigationDrawer"
            />
            <v-spacer />
            <v-menu
                left
                bottom
                dark
                nudge-bottom="40"
            >
                <template #activator="{ on }">
                    <v-btn icon v-on="on">
                        <v-icon color="white">
                            mdi-page-next
                        </v-icon>
                    </v-btn>
                </template>

                <v-list dense>
                    <v-list-item
                        v-for="(link, i) in pageMenuLinks"
                        :key="i"
                        :to="link.to"
                    >
                        <v-list-item-content>
                            <v-list-item-title v-text="link.text" />
                        </v-list-item-content>
                        <v-list-item-icon>
                            <v-icon>
                                {{ link.icon }}
                            </v-icon>
                        </v-list-item-icon>
                    </v-list-item>
                </v-list>
            </v-menu>
            <v-menu
                left
                bottom
                dark
                nudge-bottom="40"
            >
                <template #activator="{ on }">
                    <v-btn icon v-on="on">
                        <v-icon color="white">
                            mdi-account-circle
                        </v-icon>
                    </v-btn>
                </template>

                <v-list dense>
                    <v-list-item
                        v-for="(link, i) in accessibleUserLinks"
                        :key="i"
                        :to="link.to"
                    >
                        <v-list-item-content>
                            <v-list-item-title v-text="link.text" />
                        </v-list-item-content>
                        <v-list-item-icon>
                            <v-icon>
                                {{ link.icon }}
                            </v-icon>
                        </v-list-item-icon>
                    </v-list-item>
                </v-list>
            </v-menu>
        </v-app-bar>
    </header>
</template>

<script lang="ts">
import { Component, Vue, State } from 'nuxt-property-decorator'
import { Linkout } from '@/utilities/types'

@Component
export default class TheAppBar extends Vue {
    @State private navItems!: object[];
    private clipped: boolean = false;
    private userMenuLinks: Linkout[] = [
        { to: '/admin/user/changemypassword', icon: 'mdi-lock-reset', title: 'Change Your Password', text: 'Change Your Password', accessLevel: 'observer' },
        { to: '/admin', icon: 'mdi-lock', title: 'View Admin Page', text: 'Admin Tools', accessLevel: 'admin' },
        { to: '/login', icon: 'mdi-logout', title: 'Logout', text: 'Logout', accessLevel: 'observer' }
    ]

    private pageMenuLinks: Linkout[] = [
        { to: '/', icon: 'mdi-view-dashboard', title: 'View Dashboard', text: 'Dashboard' },
        { to: '/test', icon: 'mdi-star', title: 'View Test Page', text: 'Test' }
    ]

    get accessibleUserLinks () {
        return this.userMenuLinks.filter(link => this.isPermitted(link.accessLevel))
    }

    private isPermitted (accessLevel: string | undefined): boolean {
        if (accessLevel === undefined) { return false }
        return this.$permissions.isPermitted(accessLevel)
    }

    private toggleNavigationDrawer () {
        this.$nuxt.$emit('toggleDrawerVariant')
    }
}
</script>

<style lang="scss" scoped>
    ::v-deep .v-toolbar__image .v-image__image--cover {
        background-size: auto !important;
    }

    .appBarTitle {
        width: 300px;
    }
</style>
