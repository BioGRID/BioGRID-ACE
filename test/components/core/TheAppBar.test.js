
// LIBRARIES
import Vuetify from 'vuetify'
import Vuex from 'vuex'

// COMPONENTS
import TheAppBar from '@/components/core/TheAppBar.vue'

// UTILITIES
import { shallowMount, createLocalVue } from '@vue/test-utils'

// SETUP LOCALIZED VUE
const localVue = createLocalVue()
localVue.use(Vuex)

// SETUP MOCKS
const $nuxt = { $emit: jest.fn() }
const mocks = {
    $nuxt
}

let wrapper = null

// RUN BEFORE EACH TEST
beforeEach(() => {
    const vuetify = new Vuetify()
    const store = new Vuex.Store({
        state: {
            navItems: []
        }
    })
    wrapper = shallowMount(TheAppBar, {
        mocks,
        localVue,
        vuetify,
        store
    })
})

// RUN AFTER EACH TEST
afterEach(() => {
    wrapper.destroy()
})

describe('components/core/TheAppBar', () => {
    test('renders', () => {
        expect(wrapper.exists()).toBe(true)
    })
    test('renders properly', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })
    test('app bar nav icon click fires an emit event', async () => {
        await wrapper.find('.navDrawerIcon').trigger('click')
        expect($nuxt.$emit).toHaveBeenCalledWith('toggleDrawerVariant')
    })
})
