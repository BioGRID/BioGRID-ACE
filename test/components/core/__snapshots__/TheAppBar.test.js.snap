// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/core/TheAppBar renders properly 1`] = `
"<header class=\\"the-app-bar\\">
  <v-app-bar-stub color=\\"primary\\" elevation=\\"0\\" tag=\\"header\\" extensionheight=\\"48\\" src=\\"\\" value=\\"true\\" app=\\"true\\" fadeimgonscroll=\\"true\\" hideonscroll=\\"true\\" class=\\"black--text\\">
    <v-app-bar-nav-icon-stub color=\\"tertiary\\" class=\\"mr-3 navDrawerIcon\\"></v-app-bar-nav-icon-stub>
    <v-spacer-stub></v-spacer-stub>
    <v-menu-stub opendelay=\\"0\\" closedelay=\\"0\\" contentclass=\\"\\" bottom=\\"true\\" left=\\"true\\" dark=\\"true\\" maxwidth=\\"auto\\" nudgebottom=\\"40\\" nudgeleft=\\"0\\" nudgeright=\\"0\\" nudgetop=\\"0\\" nudgewidth=\\"0\\" openonclick=\\"true\\" closeonclick=\\"true\\" closeoncontentclick=\\"true\\" maxheight=\\"auto\\" origin=\\"top left\\" transition=\\"v-menu-transition\\">
      <v-list-stub tag=\\"div\\" dense=\\"true\\">
        <v-list-item-stub activeclass=\\"\\" to=\\"/\\" tag=\\"div\\">
          <v-list-item-content-stub>
            <v-list-item-title-stub>Dashboard</v-list-item-title-stub>
          </v-list-item-content-stub>
          <v-list-item-icon-stub>
            <v-icon-stub>mdi-view-dashboard</v-icon-stub>
          </v-list-item-icon-stub>
        </v-list-item-stub>
        <v-list-item-stub activeclass=\\"\\" to=\\"/test\\" tag=\\"div\\">
          <v-list-item-content-stub>
            <v-list-item-title-stub>Test</v-list-item-title-stub>
          </v-list-item-content-stub>
          <v-list-item-icon-stub>
            <v-icon-stub>mdi-star</v-icon-stub>
          </v-list-item-icon-stub>
        </v-list-item-stub>
        <v-list-item-stub activeclass=\\"\\" to=\\"/test/admin\\" tag=\\"div\\">
          <v-list-item-content-stub>
            <v-list-item-title-stub>Admin Test</v-list-item-title-stub>
          </v-list-item-content-stub>
          <v-list-item-icon-stub>
            <v-icon-stub>mdi-head-minus</v-icon-stub>
          </v-list-item-icon-stub>
        </v-list-item-stub>
      </v-list-stub>
    </v-menu-stub>
    <v-menu-stub opendelay=\\"0\\" closedelay=\\"0\\" contentclass=\\"\\" bottom=\\"true\\" left=\\"true\\" dark=\\"true\\" maxwidth=\\"auto\\" nudgebottom=\\"40\\" nudgeleft=\\"0\\" nudgeright=\\"0\\" nudgetop=\\"0\\" nudgewidth=\\"0\\" openonclick=\\"true\\" closeonclick=\\"true\\" closeoncontentclick=\\"true\\" maxheight=\\"auto\\" origin=\\"top left\\" transition=\\"v-menu-transition\\">
      <v-list-stub tag=\\"div\\" dense=\\"true\\">
        <v-list-item-stub activeclass=\\"\\" to=\\"/login\\" tag=\\"div\\">
          <v-list-item-content-stub>
            <v-list-item-title-stub>Logout</v-list-item-title-stub>
          </v-list-item-content-stub>
          <v-list-item-icon-stub>
            <v-icon-stub>mdi-logout</v-icon-stub>
          </v-list-item-icon-stub>
        </v-list-item-stub>
      </v-list-stub>
    </v-menu-stub>
  </v-app-bar-stub>
</header>"
`;
